saleor/graphql/product/bulk_mutations/products.py

```diff
@@ -71,6 +90,188 @@ class Meta:
+            try:
+                cleaned_input["attributes"] = cls.clean_attributes(
+                    attributes, data["product_type"]
+                )
+            except ValidationError as exc:
+                raise ValidationError({"attributes": exc})
+
+        return cleaned_input
+
+    @classmethod
+    def add_indexes_to_errors(cls, index, error, error_dict):
+        for key, value in error.error_dict.items():
+            for e in value:
+                if e.params:
+                    e.params["index"] = index
+                else:
+                    e.params = {"index": index}
+            error_dict[key].extend(value)
+
+    @classmethod
+    def save(cls, info, instance, cleaned_input):
+        instance.save()
+        # Recalculate the "minimal variant price" for the parent product
+        update_product_minimal_variant_price_task.delay(instance.product_id)
+
+        attributes = cleaned_input.get("attributes")
+        if attributes:
+            AttributeAssignmentMixin.save(instance, attributes)
+            instance.name = generate_name_for_variant(instance)
+            instance.save(update_fields=["name"])
+
+    @classmethod
+    def create_variants(cls, info, cleaned_inputs, product, errors):
+        instances = []
+        for index, cleaned_input in enumerate(cleaned_inputs):
+            if not cleaned_input:
+                continue
+            try:
+                instance = models.ProductVariant()
+                cleaned_input["product"] = product
+                instance = cls.construct_instance(instance, cleaned_input)
+                cls.clean_instance(instance)
+                instances.append(instance)
+            except ValidationError as exc:
+                cls.add_indexes_to_errors(index, exc, errors)
+        return instances
+
+    @classmethod
+    def clean_variants(cls, info, variants, product_type, errors):
+        cleaned_inputs = []
+        sku_list = []
+        for index, variant_data in enumerate(variants):
+            cleaned_input = None
+            try:
+                variant_data["product_type"] = product_type
+                cleaned_input = cls.clean_input(info, None, variant_data)
+            except ValidationError as exc:
+                cls.add_indexes_to_errors(index, exc, errors)
+            cleaned_inputs.append(cleaned_input if cleaned_input else None)
+
+            # Find duplicated sku in variants
+            if not variant_data.sku:
+                continue
+            if variant_data.sku in sku_list:
+                errors["sku"].append(
+                    ValidationError(
+                        "Duplicated SKU.",
+                        ProductErrorCode.UNIQUE,
+                        params={"index": index},
+                    )
+                )
+            sku_list.append(variant_data.sku)
+        return cleaned_inputs
+
+    @classmethod
+    @transaction.atomic
+    def save_variants(cls, info, instances, cleaned_inputs):
```

Comment:
Maybe we could add `assert len(instances) == len(cleaned_inputs)` just to be safe (can be safely ignored in production, e.g. compiling the python code into optimized code, which ignores the `assert`s).